apps/roam/src/utils/conceptConversion.ts

```diff
@@ -0,0 +1,184 @@
+  };
+}
+
+export function discourseRelationDataToLocalConcept(
+  context: SupabaseContext,
+  relationSchemaUid: string,
+  relationNodes: { [role: string]: string },
+): LocalConceptDataInput {
+  const roamRelation = getDiscourseRelations().find(r => r.id === relationSchemaUid);
+  if (roamRelation === undefined) {
+    throw new Error(`Invalid roam relation id ${relationSchemaUid}`)
+  }
+  const relation = discourseRelationSchemaToLocalConcept(context, roamRelation);
+  const lit_content = ((relation.literal_content)? relation.literal_content: {}) as unknown as {[key: string]: any};
+  const roles = (lit_content['roles'] as string[] | undefined)  || STANDARD_ROLES;
+  const casting = Object.fromEntries(
+    roles
+      .map((role) => [role, relationNodes[role]])
+      .filter(([, uid]) => uid !== undefined)
+  );
+  if (Object.keys(casting).length === 0) {
+    throw new Error(`No valid node UIDs supplied for roles ${roles.join(", ")}`);
+  }
+  // TODO: Also get the nodes from the representation, using QueryBuilder. That will likely give me the relation object
+  const nodeData = Object.values(casting).map((v) => getNodeExtraData(v));
+  // roundabout way to do a max from stringified dates
+  const last_modified = new Date(Math.max(...nodeData.map((nd) => new Date(nd.last_modified).getTime()))).toISOString();
+  // creation is actually creation of the relation node, not the rest of the cast, but this will do as a first approximation.
+  // Still using max, since the relation cannot be created before its cast
+  const created = new Date(Math.max(...nodeData.map((nd) => new Date(nd.created).getTime()))).toISOString();
+  const author_local_id: string = nodeData[0].author_uid;  // take any one; again until I get the relation object
+  const represented_by_local_id = (casting["target"] || Object.values(casting)[0]); // This one is tricky. Prefer the target for now.
+  return {
+    space_id: context.spaceId,
+    represented_by_local_id,
+    author_local_id,
+    created,
+    last_modified,
+    name: `${relationSchemaUid}-${Object.values(casting).join('-')}`,
+    is_schema: false,
+    schema_represented_by_local_id: relationSchemaUid,
+    local_reference_content: casting,
+  };
+}
+
+const relatedConcepts = (concept: LocalConceptDataInput): string[] => {
+  const relations = Object.values(concept.local_reference_content || {}).flat();
+  if (concept.schema_represented_by_local_id) {
+    relations.push(concept.schema_represented_by_local_id)
+  }
+  // remove duplicates
+  return [...new Set(relations)];
+}
+
+const contentRelatedToConcept = (concept: LocalConceptDataInput): string[] => {
+  const relations = relatedConcepts(concept);
+  if (concept.represented_by_local_id && !relations.includes(concept.represented_by_local_id)) {
+    relations.push(concept.represented_by_local_id)
+  }
+  return relations;
+}
+
+// From those two, a function could in theory ensure that a concept's dependent concepts are already in the database.
```

Comment:
For what two?

Reply:
Those two functions. I decided one was less likely to be used, and that the usefulness of the other would be in the context of another function which I then provided. I hope that clarifies.